sklearn/impute/_base.py

```diff
@@ -64,7 +64,56 @@ def _most_frequent(array, extra_value, n_repeat):
             return extra_value
 
 
-class SimpleImputer(TransformerMixin, BaseEstimator):
+class _BaseImputer(TransformerMixin, BaseEstimator):
+    """Base class for all imputers.
+
+    It add automatically support for `add_indicator`.
+    """
+
+    def __init__(self, missing_values=np.nan, add_indicator=False):
+        self.missing_values = missing_values
+        self.add_indicator = add_indicator
+
+    def _fit_indicator(self, X):
+        """Fit a MissingIndicator and trigger a copy of X as well."""
+        if self.add_indicator:
+            self.indicator_ = MissingIndicator(
+                missing_values=self.missing_values, error_on_new=False
+            )
+            self.indicator_.fit(X)
+        else:
+            self.indicator_ = None
+
+    def _transform_indicator(self, X):
+        """Compute the indicator mask."""
+        if self.add_indicator:
+            if not hasattr(self, 'indicator_'):
+                raise ValueError(
+                    "Make sure to call _fit_indicator before "
+                    "_transform_indicator"
+                )
+            return self.indicator_.transform(X)
+
+    def _concatenate_indicator(self, X_imputed, X_indicator):
```

Comment:
I think we can combine `_transform_indicator` and `_concatenate_indicator` into something like this:

```py
def _transform_with_indicator(self, X_imputed):
    if not self.add_indicator:
        return X_imputed
    
    # check indicator_ for developers
    if not hasattr(self, ...):
        ...

    # transform and concat
    X_indicator = self.indicator_.transform(X)

    # hstack
    hstack = sparse.hstack ...

	return hstack((X_imputed, X_indicator))
```

Reply:
No in case that we want to do the operation inplace if I recall correctly.

Reply:
Where do we plan to do this in place?

Reply:
I agree it seems like we could combine both methods? The new one would use both `X` and `X_imputed`

Reply:
to do this line:

```
X_indicator = self.indicator_.transform(X)
```

you need the original `X`. In case `copy=False` and you can do inplace imputation like in `SimpleImputer`, at the moment of the concatenation, you will not be able to compute `X_indicator` because `X` is also `X_imputed` and there no original `X` anywhere.

Does it make sense?